plutus-core/generators/Language/PlutusCore/Generators/NEAT/Type.hs

```diff
@@ -46,213 +62,374 @@ data TypeBuiltinG
 
 emptyKCS :: KCS Z
 emptyKCS = KCS{ kindOf = fromZ }
 
-extendKCS :: forall n. Kind () -> KCS n -> KCS (S n)
-extendKCS k KCS{..} = KCS{ kindOf = kindOf' }
+extKCS :: forall tyname. Kind () -> KCS tyname -> KCS (S tyname)
+extKCS k KCS{..} = KCS{ kindOf = kindOf' }
   where
-    kindOf' :: S n -> Kind ()
+    kindOf' :: S tyname -> Kind ()
     kindOf' FZ     = k
     kindOf' (FS i) = kindOf i
 
 
--- * Normalising
-
--- |Extend renamings.
-extendRen :: (n -> m) -> S n -> S m
-extendRen _ FZ     = FZ
-extendRen r (FS i) = FS (r i)
-
--- |Simultaneous renaming of variables in generated types.
-renameTypeG :: (n -> m) -> TypeG n -> TypeG m
-renameTypeG r (TyVarG i)             = TyVarG (r i)
-renameTypeG r (TyFunG ty1 ty2)       = TyFunG (renameTypeG r ty1) (renameTypeG r ty2)
-renameTypeG r (TyIFixG ty1 k ty2)    = TyIFixG (renameTypeG r ty1) k (renameTypeG r ty2)
-renameTypeG r (TyForallG k ty)       = TyForallG k (renameTypeG (extendRen r) ty)
-renameTypeG _ (TyBuiltinG tyBuiltin) = TyBuiltinG tyBuiltin
-renameTypeG r (TyLamG ty)            = TyLamG (renameTypeG (extendRen r) ty)
-renameTypeG r (TyAppG ty1 ty2 k)     = TyAppG (renameTypeG r ty1) (renameTypeG r ty2) k
-
--- |Extend substitutions.
-extendSub :: (n -> TypeG m) -> S n -> TypeG (S m)
-extendSub _ FZ     = TyVarG FZ
-extendSub s (FS i) = renameTypeG FS (s i)
-
--- |Simultaneous substitution of variables in generated types.
-substTypeG :: (n -> TypeG m) -> TypeG n -> TypeG m
-substTypeG s (TyVarG i)             = s i
-substTypeG s (TyFunG ty1 ty2)       = TyFunG (substTypeG s ty1) (substTypeG s ty2)
-substTypeG s (TyIFixG ty1 k ty2)    = TyIFixG (substTypeG s ty1) k (substTypeG s ty2)
-substTypeG s (TyForallG k ty)       = TyForallG k (substTypeG (extendSub s) ty)
-substTypeG _ (TyBuiltinG tyBuiltin) = TyBuiltinG tyBuiltin
-substTypeG s (TyLamG ty)            = TyLamG (substTypeG (extendSub s) ty)
-substTypeG s (TyAppG ty1 ty2 k)     = TyAppG (substTypeG s ty1) (substTypeG s ty2) k
+-- ** Type checking
+
+checkTypeG
+  :: Eq tyname
+  => KCS tyname
+  -> TCS tyname name
+  -> TypeG tyname
+  -> TermG tyname name
+  -> Cool
+checkTypeG _ tcs ty (VarG i)
+  = varTypeOk
+  where
+    varTypeOk = toCool $ ty == typeOf tcs i
+
+checkTypeG kcs tcs (TyForallG k ty) (TyAbsG tm)
+  = tmTypeOk
+  where
+    tmTypeOk = checkTypeG (extKCS k kcs) (firstTCS FS tcs) ty tm
+
+checkTypeG kcs tcs (TyFunG ty1 ty2) (LamAbsG tm)
+  = tyKindOk &&& tmTypeOk
+  where
+    tyKindOk = checkKindG kcs (Type ()) ty1
+    tmTypeOk = checkTypeG kcs (extTCS ty1 tcs) ty2 tm
+
+checkTypeG kcs tcs ty2 (ApplyG tm1 tm2 ty1)
+  = tm1TypeOk &&& tm2TypeOk
+  where
+    tm1TypeOk = checkTypeG kcs tcs (TyFunG ty1 ty2) tm1
+    tm2TypeOk = checkTypeG kcs tcs ty1 tm2
+
+checkTypeG kcs tcs vTy (TyInstG tm vCod ty k)
+  = tmTypeOk &&& tyKindOk &&& tyOk
+  where
+    tmTypeOk = checkTypeG kcs tcs (TyForallG k vCod) tm
+    tyKindOk = checkKindG kcs k ty
+    tyOk = vTy == normalizeTypeG (TyAppG (TyLamG vCod) ty k)
```

Comment:
... although we check that here, so perhaps not.

Reply:
agreed.